geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/KafkaConsumerFeatureSource.scala

```diff
@@ -0,0 +1,203 @@
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter.expression.{Literal, PropertyName}
+import org.opengis.filter.identity.FeatureId
+import org.opengis.filter.spatial.{BBOX, BinarySpatialOperator, Within}
+import org.opengis.filter.{Filter, IncludeFilter, Or}
+
+import scala.collection.JavaConversions._
+
+class KafkaConsumerFeatureSource(entry: ContentEntry,
+                                 schema: SimpleFeatureType,
+                                 eb: EventBus,
+                                 producer: KafkaFeatureConsumer,
+                                 query: Query)
+  extends ContentFeatureStore(entry, query) {
+
+  type FR = FeatureReader[SimpleFeatureType, SimpleFeature]
+  var qt = new Quadtree
+
+  case class FeatureHolder(sf: SimpleFeature, env: Envelope) {
+    override def hashCode(): Int = sf.hashCode()
+
+    override def equals(obj: scala.Any): Boolean = obj match {
+      case other: FeatureHolder => sf.equals(other.sf)
+      case _ => false
+    }
+  }
+
+  val features = scala.collection.mutable.HashMap.empty[String, FeatureHolder]
+
+  eb.register(this)
+
+  @Subscribe
+  def processProtocolMessage(msg: KafkaGeoMessage): Unit = msg match {
+    case update: CreateOrUpdate => processNewFeatures(update)
+    case del: Delete            => removeFeature(del)
+    case Clear                  => clear()
+    case _     => throw new IllegalArgumentException("Should never happen")
+  }
+
+  def processNewFeatures(update: CreateOrUpdate): Unit = {
+    val sf = update.f
+    val id = update.id
+    if(features.contains(id)) {
+      val old = features(id)
+      qt.remove(old.env, sf)
+    }
+    val env = sf.geometry.getEnvelopeInternal
+    qt.insert(env, sf)
+    features.put(sf.getID, FeatureHolder(sf, env))
+  }
+
+  def removeFeature(toDelete: Delete): Unit = {
+    features.remove(toDelete.id).foreach { holder =>
+      qt.remove(holder.env, holder.sf)
+    }
+  }
+
+  def clear(): Unit = {
+    features.clear()
+    qt = new Quadtree
+  }
+
+  override def getBoundsInternal(query: Query) =
+    ReferencedEnvelope.create(new Envelope(-180, 180, -90, 90), DefaultGeographicCRS.WGS84)
+
+  override def buildFeatureType(): SimpleFeatureType = schema
+
+  override def getCountInternal(query: Query): Int =
+    getReaderInternal(query).getIterator.size
+
+  override def getReaderInternal(query: Query): FR = getReaderForFilter(query.getFilter)
+
+  def getReaderForFilter(f: Filter): FR =
+    f match {
+      case o: Or             => or(o)
+      case i: IncludeFilter  => include(i)
+      case w: Within         => within(w)
+      case b: BBOX           => bbox(b)
+      case id: FidFilterImpl => fid(id)
+    }
+
+  type DFR = DelegateFeatureReader[SimpleFeatureType, SimpleFeature]
+  type DFI = DelegateFeatureIterator[SimpleFeature]
+
+  def include(i: IncludeFilter) = new DFR(schema, new DFI(features.values.map(_.sf).iterator))
```

Comment:
Looks like values just calls valuesIterator ultimately.  I just usually use the iterator one cause I thought values might make some other collection.